src/createAsyncThunk.ts

```diff
@@ -0,0 +1,183 @@
+import { Dispatch } from 'redux'
+import nanoid from 'nanoid'
+import { createAction } from './createAction'
+
+export type BaseThunkAPI<S, E, D extends Dispatch = Dispatch> = {
+  dispatch: D
+  getState: () => S
+  extra: E
+  requestId: string
+  signal: AbortSignal
+}
+
+/**
+ * @alpha
+ */
+export interface SerializedError {
+  name?: string
+  message?: string
+  stack?: string
+  code?: string
+}
+
+const commonProperties: (keyof SerializedError)[] = [
+  'name',
+  'message',
+  'stack',
+  'code'
+]
+
+// Reworked from https://github.com/sindresorhus/serialize-error
+export const miniSerializeError = (value: any): any => {
+  if (typeof value === 'object' && value !== null) {
+    const simpleError: SerializedError = {}
+    for (const property of commonProperties) {
+      if (typeof value[property] === 'string') {
+        simpleError[property] = value[property]
+      }
+    }
+
+    return simpleError
+  }
+
+  return value
+}
+
+/**
+ *
+ * @param type
+ * @param payloadCreator
+ *
+ * @alpha
+ */
+export function createAsyncThunk<
+  Returned,
+  ThunkArg = void,
+  State = unknown,
+  Extra = unknown,
+  DispatchType extends Dispatch = Dispatch,
+  ActionType extends string = string,
+  ThunkAPI extends BaseThunkAPI<any, any, any> = BaseThunkAPI<
+    State,
+    Extra,
+    DispatchType
+  >
+>(
+  type: ActionType,
+  payloadCreator: (
```

Comment:
Could these generics be passed in through this `payloadCreator` method? 

Maybe something this could work:

```ts
type payloadCreator<Returned, ThunkArg = void, State = unknown, Extra = unknown, DispatchType = unknown>(
  arg: ThunkArg,
  thunkAPI: BaseThunkAPI<State, Extra, DispatchType>
) => Promise<Returned> | Returned

export function createAsyncThunk<
  Returned,
  ThunkArg = void,
  State = unknown,
  Extra = unknown,
  DispatchType = unknown
>(
  type: ActionType,
  payloadCreator: payloadCreator<Returned, ThunkArg, State, Extra, DispatchType>
) { ... }

This can help reduce how many generics you have in the createAsyncThunk definition

Reply:
Just tried playing with this, but it's 1 AM and my head hurts.  Gonna go to bed now :)

@phryneas , if you've got time tomorrow, please give this a shot and see what happens.

Reply:
I'll take a look. I had hit similar stopgaps (inference around parameters of functions that are represented by an inferred generic seems to be pretty weak), but I was hurrying because I wanted to give you that worked at all.
Now I've got some extra time & can take a thorough look.

But there's a more pressing matter before that: the generated types are plain wrong. This is a known bug in api-extractor, I know a workaround & will apply it.